Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/ObjCStubs.kt

```diff
@@ -446,139 +423,150 @@ abstract class ObjCContainerStub(stubGenerator: StubGenerator,
-                    // TODO: handle the case when protocol getter stub can't be compiled.
-                    genProtocolGetter(stubGenerator, nativeBacked, container)
-                }
-
-                externalObjCClassAnnotationName.applyToStrings(protocolGetter)
-            }
-            is ObjCClass -> {
-                protocolGetter = null
-                val binaryName = container.binaryName
-                if (binaryName != null) {
-                    externalObjCClassAnnotationName.applyToStrings("", binaryName)
-                } else {
-                    externalObjCClassAnnotationName
-                }
-            }
-        }
-
-        this.classHeader = "$externalObjCClassAnnotation $keywords $name : $supersString"
+        interfaces
     }
 
-    open fun generateBody(context: StubGenerationContext): Sequence<String> {
-        var result = (propertyStubs.asSequence() + methodStubs.asSequence())
-                .flatMap { sequenceOf("") + it.generate(context) }
-
-        if (container is ObjCClass && methodStubs.none {
-            it.method.isInit && it.method.parameters.isEmpty() && context.nativeBridges.isSupported(it)
-        }) {
+    protected fun buildBody(): Pair<List<PropertyStub>, List<FunctionalStub>> {
+        // TODO: check `context.nativeBridges.isSupported(it)`
+        val defaultConstructor =  if (container is ObjCClass && methodToStub.values.none { it.isDefaultConstructor() }) {
             // Always generate default constructor.
             // If it is not produced for an init method, then include it manually:
-            result += sequenceOf("", "protected constructor() {}")
-        }
+            ConstructorStub(listOf(), listOf(), VisibilityModifier.PROTECTED)
+        } else null
 
-        return result
+        return Pair(
+                propertyBuilders.flatMap { it.build() },
+                methodToStub.values.flatMap { it.build() } + listOfNotNull(defaultConstructor)
+        )
     }
-
-    override fun generate(context: StubGenerationContext): Sequence<String> = block(classHeader, generateBody(context))
 }
 
-open class ObjCClassOrProtocolStub(
-        stubGenerator: StubGenerator,
+internal sealed class ObjCClassOrProtocolStubBuilder(
+        context: StubsBuildingContext,
         private val container: ObjCClassOrProtocol
-) : ObjCContainerStub(
-        stubGenerator,
+) : ObjCContainerStubBuilder(
+        context,
         container,
-        metaContainerStub = object : ObjCContainerStub(stubGenerator, container, metaContainerStub = null) {}
-) {
-    override fun generate(context: StubGenerationContext) =
-            metaContainerStub!!.generate(context) + "" + super.generate(context)
+        metaContainerStub = object : ObjCContainerStubBuilder(context, container, metaContainerStub = null) {
+
+            override fun build(): List<StubIrElement> {
+                val (properties, methods) = buildBody()
```

Comment:
This `buildBody` + `ClassStub.Simple()` combination looks nearly the same here and in next two classes.
Isn't it possible to extract the common logic?